ktor-features/ktor-mustache/test/io/ktor/mustache/MustacheTest.kt

```diff
@@ -0,0 +1,161 @@
+
+                val contentTypeText = assertNotNull(response.headers[HttpHeaders.ContentType])
+                assertEquals(ContentType.Text.Html.withCharset(Charsets.UTF_8), ContentType.parse(contentTypeText))
+            }
+        }
+    }
+
+    @Test
+    fun `Fill template and expect eTag set when it is provided`() {
+        withTestApplication {
+            application.setupMustache()
+            application.install(ConditionalHeaders)
+
+            application.routing {
+                get("/") {
+                    call.respond(MustacheContent(TemplateWithPlaceholder, DefaultModel, "e"))
+                }
+            }
+
+            assertEquals("e", handleRequest(HttpMethod.Get, "/").response.headers[HttpHeaders.ETag])
+        }
+    }
+
+
+    @Test
+    fun `Render empty model`() {
+        withTestApplication {
+            application.setupMustache()
+            application.install(ConditionalHeaders)
+
+            application.routing {
+                get("/") {
+                    call.respond(MustacheContent(TemplateWithoutPlaceholder, null, "e"))
+                }
+            }
+
+            handleRequest(HttpMethod.Get, "/").response.let { response ->
+
+                val lines = response.content!!.lines()
+
+                assertEquals("<p>Hello, Anonymous</p>", lines[0])
+                assertEquals("<h1>Hi!</h1>", lines[1])
+            }
+        }
+    }
+
+    @Test
+    fun `Render template compressed with GZIP`() {
+        withTestApplication {
+            application.setupMustache()
+            application.install(Compression)
+            application.install(ConditionalHeaders)
+
+            application.routing {
+                get("/") {
+                    call.respondTemplate(TemplateWithPlaceholder, DefaultModel, "e")
+                }
+            }
+
+            handleRequest(HttpMethod.Get, "/") {
+                addHeader(HttpHeaders.AcceptEncoding, "gzip")
+            }.response.let { response ->
+                val content = GZIPInputStream(response.byteContent!!.inputStream()).reader().readText()
+
+                val lines = content.lines()
+
+                assertEquals("<p>Hello, 1</p>", lines[0])
+                assertEquals("<h1>Hello World!</h1>", lines[1])
+            }
+        }
+    }
+
+    @Test
+    fun `Render template without eTag`() {
+        withTestApplication {
+            application.setupMustache()
+            application.install(ConditionalHeaders)
+
+            application.routing {
+
+                get("/") {
+                    call.respond(MustacheContent(TemplateWithPlaceholder, DefaultModel))
+                }
+            }
+
+            assertEquals(null, handleRequest(HttpMethod.Get, "/").response.headers[HttpHeaders.ETag])
+        }
+    }
+
+    private fun Application.setupMustache() {
+        install(Mustache) {
+            DefaultMustacheFactory()
```

Comment:
This seems to be redundant isn't it?

Reply:
Yeah that could be possible. I'll have a look on it

Reply:
You were right, I implemented the feature configuration wrong there. 
Now I added an explicit configuration, so the usage of other Factories is possible, but default is DefaultMustacheFactory